framework/src/userobjects/ModifyElementSubdomainByVar.C

```diff
@@ -0,0 +1,59 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ModifyElementSubdomainByVar.h"
+
+registerMooseObject("MooseApp", ModifyElementSubdomainByVar);
+
+InputParameters
+ModifyElementSubdomainByVar::validParams()
+{
+  InputParameters params = ElementSubdomainModifier::validParams();
+
+  params.addRequiredCoupledVar(
+      "coupled_var", "Coupled variable whose value is used to calculate the desired subdomain ID.");
+  params.addClassDescription(
+      "Modify the element subdomain ID based on the provided variable value.");
+  return params;
+}
+
+ModifyElementSubdomainByVar::ModifyElementSubdomainByVar(const InputParameters & parameters)
+  : ElementSubdomainModifier(parameters), _v(coupledValue("coupled_var"))
+{
+}
+
+SubdomainID
+ModifyElementSubdomainByVar::computeSubdomainID()
+{
+  // Calculate the desired subdomain ID for the current element.
+  Real val = 0.0;
+  for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
+    val += _v[qp] * _JxW[qp] * _coord[qp];
+  val /= _current_elem_volume;
+
+  SubdomainID sid = (unsigned int)(round(val));
+
+  // Check if the desired subdomain ID exists in the mesh. Use the nearest subdomain ID if the
+  // desired subdomain ID does not exist in the mesh.
+  if (_mesh.meshSubdomains().find(sid) == _mesh.meshSubdomains().end())
+  {
+    auto it = _mesh.meshSubdomains().lower_bound(sid);
+    SubdomainID nearest_id =
+        it == _mesh.meshSubdomains().end() ? *_mesh.meshSubdomains().rbegin() : *it;
+    mooseDoOnce(mooseWarning(
```

Comment:
We dont like to do this anymore.

I suggest you keep a vector (with a lock for threads?) that keeps track of which subdomains you have output a warning for, to output the warning only once per subdomain